src/crypto/crypto.cpp

```diff
@@ -221,7 +220,10 @@ namespace crypto {
 #endif
     buf.h = prefix_hash;
     buf.key = pub;
-    random_scalar(k);
+    {
+      boost::lock_guard<boost::mutex> lock(random_lock);
+      random_scalar(k);
+    }
```

Comment:
I'd move the lock to a new random_scalar function, and rename the current random_scalar to random_scalar_not_thread_safe, since it is not, and that was a source of bugs earlier.